doniclient/osc/common.py

```diff
@@ -0,0 +1,128 @@
+import argparse
+import json
+import logging
+from argparse import FileType, Namespace
+
+import yaml
+from cliff.columns import FormattableColumn
+from keystoneauth1.exceptions import HttpError
+from osc_lib import utils as osc_utils
+from osc_lib.command import command
+
+LOG = logging.getLogger(__name__)  # Get the logger of this module
+
+
+class OutputFormat:
+    columns = (
+        "uuid",
+        "name",
+        "project_id",
+        "hardware_type",
+        "properties",
+    )
+
+
+class YamlColumn(FormattableColumn):
+    def human_readable(self):
+        return yaml.dump(self._value)
+
+
+class HardwareSerializer(object):
+    def serialize_hardware(self, hw_dict: "dict", columns: "list[str]"):
+        return osc_utils.get_dict_properties(
+            hw_dict,
+            columns,
+            formatters={
+                "properties": YamlColumn,
+                "workers": YamlColumn,
+            },
+        )
+
+
+class GroupedAction(argparse.Action):
+    """Set property based on dest.key."""
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        group, dest = self.dest.split(".", 2)
+
+        try:
+            group_dict = getattr(namespace, group)
+        except AttributeError:
+            group_dict = {}
+        group_dict.update({dest: values})
+        setattr(namespace, group, group_dict)
```

Comment:
If you do `setdefault` you don't need this part, b/c it's already set on the object.